### ISSUE TITLE
The error occurred while executing the backward method using the LinkAsTorchModel class

### ISSUE THREAD

**Issue by take0212** [chainer/chainer-pytorch-migration#23]
I am trying to call a chainer model using the LinkAsTorchModel class.
However, I am getting the following error while executing the backward method.
Can you give me any advice?

![image](https://user-images.githubusercontent.com/43694878/81499932-c5c17480-9309-11ea-9083-755f4c039de4.png)

I want to call a chainer model from a pytorch trainer.
I refer to the following site to implement the code.
https://chainer.github.io/migration-guide/
But that is difficult for me.
I need more sample codes.

**Comment by emcastillo** [comment]
Can you show the code of how do you create the model and how do you call backward?

**Comment by take0212** [comment]
Thank you for your comment.
I copied my code below.

chainer device
```
            import chainer
            self.chainer_device = chainer.get_device(self.device)
 ```

device
```
        if self.device >= 0:
            self.device = 'cuda'
        else:
            self.device = 'cpu'
        self.device = torch.device(self.device)
```
MLP class
```
class MLP(chainer.Chain):

    def __init__(self, n_mid_units=100, n_out=10):
        super(MLP, self).__init__()
        with self.init_scope():
            self.l1=L.Linear(None, n_mid_units)
            self.l2=L.Linear(None, n_mid_units)
            self.l3=L.Linear(None, n_out)

    def forward(self, x):
        h1 = F.relu(self.l1(x))
        h2 = F.relu(self.l2(h1))
        return self.l3(h2)
```

The sub class of MLP and classifier
```
from myfw.mychainer.model.MLP import MLP

class MLPAndClassifier(MLP):

    def forward(self, x):
        y = super(MLPAndClassifier, self).forward(x)
        return F.log_softmax(y)
```

create model
```
            from myfw.mychainer.model.MLPAndClassifier import MLPAndClassifier
            self.model = MLPAndClassifier()
 
            self.input_size = (28 * 28,)
```

wrap  model
```
            import chainer_pytorch_migration as cpm
            self.model.to_device(self.chainer_device)
            self.model(self.chainer_device.xp.zeros((1,) + self.input_size).astype('f'))
            self.model = cpm.LinkAsTorchModel(self.model)
```

set  optimizer

create trainer
```
        self.trainer = create_supervised_trainer(self.model, self.optimizer, F.nll_loss, device=self.device)
```

set event handler

run
```
        self.trainer.run(self.train_loader, max_epochs=self.max_epochs)
```

**Comment by emcastillo** [comment]
is `F.nll_loss` a chainer or a torch function?

**Comment by emcastillo** [comment]
Also, when wrapping a chainer model with `LinkAsTorchModel` you should use a torch manual training loop instead of the chainer trainer. The optimizer needs to be a torch optimizer too.

https://towardsdatascience.com/the-pytorch-training-loop-3c645c56665a

**Comment by take0212** [comment]
import torch function at the top of the file.

```
# import
import torch
import torch.nn.functional as F
from torch.utils.data import DataLoader
from torch import optim
```

**Comment by take0212** [comment]
I have already set the torch optimizier.
```
        self.optimizer = optim.SGD(self.model.parameters(), lr=self.lr, momentum=self.momentum)
```

**Comment by emcastillo** [comment]
Can you please share a gist with the complete example so I can download it and run it straight away?
Thank you

**Comment by take0212** [comment]
I'll upload my code tonight.

**Comment by take0212** [comment]
I uploaded my codes to the github.
https://github.com/take0212/pytorch-chainer-combination
Could you please check them?

**Comment by emcastillo** [comment]
This case is the contrary
for `LinkAsTorchModel` the migration guide says
`The graph (forward/backward) must be constructed and traversed in Chainer`
This means that the loss function must be a chainer function too.

You need to pass a custom step function to your ignite trainer to deal with the loss calculation here as in chainer models the loss is the return value of the Link `forward`
https://pytorch.org/ignite/concepts.html#engine

**Comment by take0212** [comment]
Thank you for your kind support.
I will confirm it within a few days.

**Comment by take0212** [comment]
I set the chainer loss function.

```
        from chainer.functions.loss.softmax_cross_entropy import softmax_cross_entropy
        self.trainer = create_supervised_trainer(self.model, self.optimizer, softmax_cross_entropy, device=self.device)
```

I can't understand the following comment.

> You need to pass a custom step function to your ignite trainer to deal with the loss calculation here as in chainer models the loss is the return value of the Link forward

So the following error happens.

![image](https://user-images.githubusercontent.com/43694878/81995912-ab2f2880-9686-11ea-82b0-faa69292563a.png)

![image](https://user-images.githubusercontent.com/43694878/81996160-5344f180-9687-11ea-98d8-c3ea7b672a1c.png)

![image](https://user-images.githubusercontent.com/43694878/81996052-03fec100-9687-11ea-98d0-3f1f0c5012e2.png)

**Comment by take0212** [comment]
I implemented the custom process function to convert Tensor to cupy and passed it to ignite.engine.Engine.
My sample code can run without error.
But in fact, it can't be learned at all.

I updated my code to my github.
Could you check it again?

I would like to use Chainer model in PyTorch code(torchvision.datasets and ignite.engine).
Is it possible?

Should I convert torch.utils.data.DataLoader to chainer.iterators.SerialIterator instead of passing the custom process function to ignite.engine.Engine in order to convert a Tensor to a cupy?

I also imported pytorch-pfn-extras.
I use chainer pytorch_pfn_extras.training.extensions and I can evaluate the model in case of PyTorch model and PyTorch code(torchvision.datasets and ignite.engine).
I also use pytorch_pfn_extras.nn.LazyLinear.
Thank you.

**Comment by emcastillo** [comment]
Sorry, the code is difficult to follow.
Can you create a single .py file without all the boilerplate code that reproduces the error?

**Comment by take0212** [comment]
I am currently referencing the sample code below.

https://gist.github.com/emcastillo/8be64990940ce23147cc08584bea985b

**Comment by emcastillo** [comment]
The issue seems solved so let us close it.
Thanks

**Comment by take0212** [comment]
Finally, I solved the problem.

I wrote my sample code by referring to your code.

I customized pytorch trainer function and prepare_batch function so that it also works on GPU.

I uploaded the fixed code to my github.

If you have any comments, could you reply to me?
If you don't reply to me, I'll publish the sample code.

I appreciate your support so far.

**Comment by take0212** [comment]
I published it.
https://github.com/take0212/pytorch-chainer-combination
